src/com/telesign/util/TeleSignRequest.java

```diff
@@ -1,450 +1,504 @@
+	 *			[Required] A string that contains the contents of the POST
+	 *			Request's entity body.
+	 */
+	public void setPostBody(String post_body) {
+
+		body = post_body;
+	}
+
+	/**
+	 * Gets the entity body of an HTTP 1.1 POST Request message. Used for calls
+	 * to <strong>Verify.call</strong> and <strong>Verify.sms</strong>.
+	 *
+	 * @return A string that contains the contents of the POST Request's entity
+	 *		 body.
+	 */
+	public String getPostBody() {
+
+		return body;
+	}
+
+	/**
+	 * Returns the set of <em>TeleSign-specific</em> Request header fields.
+	 *
+	 * @return A sorted key/value mapping that contains the TeleSign-specific
+	 *		 Request header fields.
+	 */
+	public Map<String, String> getTsHeaders() {
+
+		return ts_headers;
+	}
+
+	/**
+	 * Returns the set of both the standard <u>and</u> the
+	 * <em>TeleSign-specific</em> Request header fields.
+	 *
+	 * @return A sorted key/value mapping that contains all of the Request header fields.
+	 */
+	public Map<String, String> getAllHeaders() {
+
+		TreeMap<String, String> tmp = new TreeMap<String, String>();
+
+		tmp.putAll(headers);
+		tmp.putAll(ts_headers);
+
+		return tmp;
+	}
+
+	/**
+	 * Returns the set of POST Request parameters.
+	 *
+	 * @return A sorted key/value mapping that contains all of the POST Request parameters.
+	 */
+	public Map<String, String> getAllParams() {
+
+		return params;
+	}
+
+	/**
+	 * Creates and sends the REST request.
+	 *
+	 * @return A string containing the TeleSign web server's Response.
+	 * @throws IOException
+	 *			 A {@link java.security.SignatureException} signals that an
+	 *			 error occurred while attempting to sign the Request.
+	 */
+	public String executeRequest() throws IOException {
+
+		String signingString = getSigningString(customer_id);
+		String signature;
+		String url_output = "";
+
+		// Create the absolute form of the resource URI, and place it in a string buffer.
+		StringBuffer full_url = new StringBuffer(base).append(resource);
+
+		if (params.size() > 0) {
+
+			full_url.append("?");
+			int i = 0;
+
+			for (String key : params.keySet()) {
+
+				if (++i != 0) {
+
+					full_url.append("&");
+				}
+
+				full_url.append(URLEncoder.encode(key, "UTF-8")).append("=").append(URLEncoder.encode(params.get(key), "UTF-8"));
+			}
+		}
+
+		url = new URL(full_url.toString());
+
+		connection = (HttpURLConnection) url.openConnection();
+		connection.setConnectTimeout(connectTimeout);
+		connection.setReadTimeout(readTimeout);
+
+		// Create the Signature using the formula: Signature = Base64(HMAC-SHA( YourTeleSignAPIKey, UTF-8-Encoding-Of( StringToSign )).
+		try {
+
+			signature = encode(signingString, secret_key);
+		}
+		catch (SignatureException e) {
+
+			System.err.println("Error signing request " + e.getMessage());
+
+			return null;
+		}
+
+		String auth_header = "TSA " + customer_id + ":" + signature;
+
+		connection = (HttpURLConnection) url.openConnection();
+		connection.setConnectTimeout(connectTimeout);

```

Comment:
We are setting the connection and read timeout out twice. Also when running the unit test that currently actually call rest.telesign.com the timeout doesn't appear to be working.